index.js

```diff
@@ -71,33 +48,25 @@ class MinQuery {
 
     this.email = options.email;
     this.projectId = options.projectId;
+    this.bearerCache = new BearerCache();
   }
 
-  _request(method, path, data) {
+  async _request(method, path, data) {
     const url = `https://www.googleapis.com/bigquery/v2/projects/${this.projectId}${path}`;
 
-    return new Promise((resolve, reject) => {
-      request({
-        method,
-        body: data,
-        json: true,
-        url,
-        jwt: {
-          email: this.email,
-          key: this.key,
-          scopes: SCOPES,
-        },
-      }, (err, res) => {
-        if (!err && res && res.body.error) {
-          err = new ProtocolError(res.body.error.message || 'API response error', res);
-        }
-
-        if (err) {
-          reject(err);
-        } else {
-          resolve(res);
-        }
-      });
+    const bearer = await getBearer({
```

Comment:
Can this result in a situation where multiple calls to `getBearer` are concurrently in flight, i.e. waiting on resolution of the same underlying `Promise` instance? Is that legal...?

Reply:
Oh yeah, that's by design! The idea is: if the cache is cold and we suddenly get 100 calls to `getBearer`, we only want to make a single network request to fetch a token, and have all callers await the resolution of that one operation. And then if the cache is warm, just return a promise that immediately resolves with the value. 

`await` is just sugar for calling `.then` on the returned promise, and basically all `.then` is doing is queueing up an array of functions to call when the promise is resolved or rejected. 

Here's a simplified example:

```js
const p = new Promise(r => setTimeout(() => r('result'), 10));

p.then(r => console.log(`I got ${r}!`));
p.then(r => console.log(`I got ${r}, too!`));
```
Which prints:

```
I got result!
I got result, too!
```

Reply:
Helpful refresher - thanks!